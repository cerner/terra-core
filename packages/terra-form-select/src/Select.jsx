import React from 'react';
import PropTypes from 'prop-types';
import Variants from './shared/_variants';
import Option from './shared/_Option';
import OptGroup from './shared/_OptGroup';
import Combobox from './Combobox';
import SingleSelect from './SingleSelect';
import MultiSelect from './MultiSelect';
import Search from './SearchSelect';
import Tag from './TagSelect';

const propTypes = {
  /**
   * Whether a clear option is available to clear the selection, will use placeholder text if provided.
   * This is not applicable to the `multiple` or `tag` variants since the selection can already be deselected using the tag.
   */
  allowClear: PropTypes.bool,
  /**
   * The dropdown menu options.
   */
  children: PropTypes.node,
  /**
   * The default selected value. Can be a string, number, or array of strings/numbers.
   */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
  /**
   * Whether the select is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Additional attributes to spread onto the dropdown. ( Style, ClassNames, etc.. )
   */
  // eslint-disable-next-line react/forbid-prop-types
  dropdownAttrs: PropTypes.object,
  /**
   * Whether the select displays as Incomplete. Use when no value has been provided. _(usage note: `required` must also be set)_.
   */
  isIncomplete: PropTypes.bool,
  /**
   * Whether the select displays as Invalid. Use when value does not meet validation pattern.
   */
  isInvalid: PropTypes.bool,
  /**
   * Ensure accessibility on touch devices. Will render the dropdown menu in
   * normal DOM flow with position absolute. By default, the menu renders in a
   * portal, which is inaccessible on touch devices. Only applicable to variants
   * that include an input (e.g. `combobox`, `multiple`, `search`, and `tag`).
   */
  isTouchAccessible: PropTypes.bool,
  /**
   * The max height of the dropdown.
   */
  maxHeight: PropTypes.number,
  /**
   * @private The maximum number of options that can be selected. A value less than 2 will be ignored.
   * Only applicable to variants allowing multiple selections (e.g.; `multiple`; `tag`).
   */
  maxSelectionCount: PropTypes.number,
  /**
   * Content to display when no results are found.
   */
  noResultContent: PropTypes.node,
  /**
   * Callback function triggered when the select loses focus. function(event)
   */
  onBlur: PropTypes.func,
  /**
   * Callback function triggered when the value changes. function(value)
   */
  onChange: PropTypes.func,
  /**
   * Callback function triggered when the select is clicked. function(event)
   */
  onClick: PropTypes.func,
  /**
   * Callback function triggered when an option is deselected. function(value)
   */
  onDeselect: PropTypes.func,
  /**
   * Callback function triggered when the select receives focus. function(event)
   */
  onFocus: PropTypes.func,
  /**
   * Callback function triggered when the search criteria changes. function(searchValue)
   */
  onSearch: PropTypes.func,
  /**
   * Callback function triggered when an option is selected. function(value)
   */
  onSelect: PropTypes.func,
  /**
   * Callback function invoked for each option on search change. function(searchValue, option)
   */
  optionFilter: PropTypes.func,
  /**
   * Placeholder text.
   */
  placeholder: PropTypes.string,
  /**
   * Whether the field is required.
   */
  required: PropTypes.bool,
  /**
   * The selected value. Can be a string, number, or array of strings/numbers.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
  /**
   * The behavior of the select. One of `default`, `combobox`, `multiple`, `tag`, or `search`.
   */
  variant: PropTypes.oneOf([
    Variants.COMBOBOX,
    Variants.DEFAULT,
    Variants.MULTIPLE,
    Variants.SEARCH,
    Variants.TAG,
  ]),
};

const defaultProps = {
  allowClear: false,
  children: undefined,
  defaultValue: undefined,
  disabled: false,
  dropdownAttrs: undefined,
  isIncomplete: false,
  isInvalid: false,
  isTouchAccessible: false,
  maxSelectionCount: undefined,
  noResultContent: undefined,
  onChange: undefined,
  onDeselect: undefined,
  onSearch: undefined,
  onSelect: undefined,
  optionFilter: undefined,
  placeholder: undefined,
  required: false,
  value: undefined,
  variant: 'default',
};

function Select(props) {
  const { variant, ...otherProps } = props;

  switch (variant) {
    case Variants.COMBOBOX: {
      const { maxSelectionCount, ...comboboxProps } = otherProps;
      return <Combobox {...comboboxProps} />;
    }
    case Variants.MULTIPLE: {
      const { allowClear, ...multipleProps } = otherProps;
      return <MultiSelect {...multipleProps} />;
    }
    case Variants.SEARCH: {
      const { maxSelectionCount, ...searchProps } = otherProps;
      return <Search {...searchProps} />;
    }
    case Variants.TAG: {
      const { noResultContent, allowClear, ...tagProps } = otherProps;
      return <Tag {...tagProps} />;
    }
    case Variants.DEFAULT:
    default: {
      const {
        isTouchAccessible,
        maxSelectionCount,
        onSearch,
        optionFilter,
        ...singleSelectProps
      } = otherProps;
      return <SingleSelect {...singleSelectProps} />;
    }
  }
}

Select.Option = Option;
Select.OptGroup = OptGroup;
Select.propTypes = propTypes;
Select.defaultProps = defaultProps;
Select.isSelect = true;

export default Select;
