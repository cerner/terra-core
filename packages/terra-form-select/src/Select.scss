:local {
  .select {
    background-color: #fff;
    background-image: var(--terra-select-focus-animation-background, none);
    background-position: 0 100%;
    background-repeat: no-repeat;
    background-size: var(--terra-select-focus-background-start, 0 0);
    border-color: var(--terra-select-input-border-color, #dedfe0);
    border-radius: 3px;
    border-style: solid;
    border-width: 1px;
    cursor: pointer;
    font-size: var(--terra-select-font-size, 1.143rem);
    line-height: var(--terra-select-line-height, 1.125);
    padding: var(--terra-select-input-padding, 0.3572rem 0.429rem);
    transition-duration: var(--terra-select-focus-transition-duration, 0s);
    transition-property: background-size;
    transition-timing-function: var(--terra-select-focus-transition-timing, ease);
    width: 100%;

    &.is-disabled {
      background-color: var(--terra-select-disabled-background-color, rgba(111, 116, 119, 0.05));
      border-color: var(--terra-select-disabled-border-color, #dedfe0);
      color: var(--terra-select-disabled-color, rgba(0, 0, 0, 0.25));
      cursor: default;

      &:focus {
        background-size: 0 0;
        border-color: var(--terra-select-disabled-border-color, #dedfe0);

        /* stylelint-disable max-nesting-depth */
        [dir=rtl] &,
        [dir=ltr] & {
          box-shadow: none;
        }
        /* stylelint-enable max-nesting-depth */
      }

      &:hover {
        border-color: var(--terra-select-disabled-border-color, #dedfe0);
      }
    }

    &:hover {
      border-color: var(--terra-select-input-hover-border-color, #dedfe0);
    }

    &:focus,
    &.open {
      background-size: var(--terra-select-focus-animation-size, 0 0);
      border-color: var(--terra-select-input-focus-border-color, #4cb2e9);
      outline: none;

      [dir=rtl] & {
        box-shadow: var(--terra-select-focus-box-shadow-rtl, 0 0 7px 4px rgba(76, 178, 233, 0.35), 0 0 1px 3px rgba(76, 178, 233, 0.5), 0 0 1px 0 #fff);
      }

      [dir=ltr] & {
        box-shadow: var(--terra-select-focus-box-shadow-ltr, 0 0 7px 4px rgba(76, 178, 233, 0.35), 0 0 1px 3px rgba(76, 178, 233, 0.5), 0 0 1px 0 #fff);
      }
    }

    &.invalid {
      background-image: var(--terra-select-invalid-focus-animation-background, none);
      border-color: var(--terra-select-invalid-border-color, #dedfe0);

      &:hover {
        border-color: var(--terra-select-invalid-hover-border-color, #dedfe0);
      }

      &.open,
      &:focus {
        background-size: var(--terra-select-invalid-focus-animation-size, 0 0);
        border-color: var(--terra-select-invalid-focus-border-color, #4cb2e9);

        /* stylelint-disable max-nesting-depth */
        [dir=rtl] & {
          box-shadow: var(--terra-select-invalid-focus-box-shadow-rtl, 0 0 7px 4px rgba(76, 178, 233, 0.35), 0 0 1px 3px rgba(76, 178, 233, 0.5), 0 0 1px 0 #fff);
        }

        [dir=ltr] & {
          box-shadow: var(--terra-select-invalid-focus-box-shadow-ltr, 0 0 7px 4px rgba(76, 178, 233, 0.35), 0 0 1px 3px rgba(76, 178, 233, 0.5), 0 0 1px 0 #fff);
        }
        /* stylelint-enable max-nesting-depth */
      }
    }
  }

  .invalid {
    background-image: var(--terra-select-invalid-focus-animation-background, none);
    border-color: var(--terra-select-invalid-border-color, #dedfe0);
  }

  .invalid .select-arrow {
    color: var(--terra-select-invalid-arrow-color, auto);
  }

  .is-disabled .select-arrow {
    color: var(--terra-select-disabled-arrow-color, #b2b5b6);
  }

  .select-arrow {
    align-self: center;
    color: var(--terra-select-arrow-color, auto);
    display: flex;
    height: var(--terra-select-arrow-height, 0.715rem);
    margin: var(--terra-select-arrow-margin, 0.286rem 0);
    width: var(--terra-select-arrow-width, 1rem);
  }

  .default-option,
  .display.default-option {
    color: var(--terra-select-default-option-color, #bcbfc0);
    font-style: var(--terra-select-default-option-font-style, normal);
  }

  .display {
    color: var(--terra-select-input-text-color, #2d3539);
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .is-disabled .display {
    color: var(--terra-select-disabled-font-color, #b2b5b6);
  }

  .select-menu {
    background-color: var(--terra-select-menu-background-color, #fff);
    border: solid 1px #b2b5b6;
    border-radius: var(--terra-select-menu-border-raidus, 0);
    box-shadow: var(--terra-select-menu-box-shadow, 0 3px 6px 0 rgba(0, 0, 0, 0.25));
    margin: 0;
    padding-bottom: var(--terra-select-menu-bottom-padding, 0.214rem);

    [dir=rtl] & {
      box-shadow: var(--terra-select-menu-box-shadow-rtl, 0 3px 6px 0 rgba(0, 0, 0, 0.25));
    }

    [dir=ltr] & {
      box-shadow: var(--terra-select-menu-box-shadow-ltr, 0 3px 6px 0 rgba(0, 0, 0, 0.25));
    }
  }

  @supports (-webkit-backdrop-filter: blur(var(--terra-select-menu-blur, 0))) {
    .select-menu {
      backdrop-filter: blur(var(--terra-select-menu-blur, 0));
      background-color: var(--terra-select-menu-blur-background-color, #fff);
    }
  }
}
