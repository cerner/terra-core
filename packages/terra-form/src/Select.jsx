import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import 'terra-base/lib/baseStyles';
import styles from './Select.scss';

const cx = classNames.bind(styles);

const propTypes = {

  /**
   * WARNING: This prop is deprecated, please use the options prop.
   * List of choices to be selected.
   * If choices and options array are supplied, options array will be used over choices.
   */
  choices: PropTypes.array,

  /**
   * List of object key and value pairs for choices to be selected.
   * If choices and options array are supplied, options array will be used over choices.
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    display: PropTypes.string.isRequired,
  })),

  /**
   * Function to trigger when the user changes the select value. Provide a function to create a controlled input.
   */
  onChange: PropTypes.func,

  /**
   * Name of the select field.
   */
  name: PropTypes.string,

  /**
   * Whether the select is required or not.
   */
  required: PropTypes.bool,

  /**
   * The value to start the select on.
   */
  defaultValue: PropTypes.string,

  /**
   * The value of the select element. Use this to create a controlled input.
   */
  value: PropTypes.string,
};

const defaultProps = {
  choices: null,
  options: null,
  onChange: undefined,
  name: null,
  required: false,
  defaultValue: undefined,
  value: undefined,
};

const Select = ({
  choices,
  options,
  onChange,
  name,
  required,
  defaultValue,
  value,
  ...customProps
}) => {
  const additionalSelectProps = Object.assign({}, customProps);
  const selectClasses = cx(['select', additionalSelectProps.className]);

  if (required) {
    additionalSelectProps['aria-required'] = 'true';
  }

  let finalOptions;
  if (choices && options === null) {
    // Build out key & value array for deprecated choices prop
    finalOptions = choices.map(choice => ({ value: choice, display: choice }));
    /* eslint-disable no-console */
    console.warn('The choices prop for the Terra Form Select component is deprecated and will be removed in a later release. Please use the options prop instead https://terra-ui.herokuapp.com/components/core/form#select.');
    /* eslint-disable no-console */
  } else {
    finalOptions = options;
  }

  return (
    <select
      {...additionalSelectProps}
      name={name}
      required={required}
      onChange={onChange}
      defaultValue={defaultValue}
      value={value}
      className={selectClasses}
    >
      {finalOptions.map(option => <option key={`${option.value}-${option.display}`} value={option.value}>{option.display}</option>)}
    </select>
  );
};

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
