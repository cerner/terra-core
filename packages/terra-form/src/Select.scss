:local {
  /* stylelint-disable selector-class-pattern */
  
  .select {
    background-color: #fff;
    border: #dedfe0 solid 1px;
    border-radius: 0.25em;
    font-size: 1em;
    height: 1.857em;
    padding: 0.178em;
    width: 100%;

    &[required] {
      box-shadow: unset;
      outline: none;
    }
  }
}
