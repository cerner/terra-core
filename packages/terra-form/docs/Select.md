# Select

Generic select component which represents an HTML select element directly.

## Usage

```jsx
import React from 'react';
import Select from 'terra-form/lib/Select';

<Select
  options={[{ value: 'm', display: 'moo' },
            { value: 'b', display: 'boo' },
            { value: 'z', display: 'zar' }]}
  name="foo"
  defaultValue="b"
  required
/>
```
