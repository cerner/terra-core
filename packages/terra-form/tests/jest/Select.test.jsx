import React from 'react';
import Select from '../../src/Select';

it('should render a default Select component', () => {
  const select = <Select options={[]} />;
  const wrapper = shallow(select);
  expect(wrapper).toMatchSnapshot();
});

it('should render a Select component with one option', () => {
  const select = <Select options={[{ value: 'm', display: 'moo' }]} />;
  const wrapper = shallow(select);
  expect(wrapper).toMatchSnapshot();
});

it('should render a Select when all the possible props are passed into it', () => {
  const select = (
    <Select
      options={[{ value: 'm', display: 'moo' }]}
      name="description"
      required
    />);

  const wrapper = shallow(select);
  expect(wrapper).toMatchSnapshot();
});

it('should render as uncontrolled when just the required fields and a default value is passed into the Select', () => {
  const select = <Select options={[{ value: 'm', display: 'moo' }, { value: 'b', display: 'boo' }]} defaultValue="b" />;
  const wrapper = mount(select);
  expect(wrapper).toMatchSnapshot();
});

it('should render as controlled when just the required fields and a onChange function is passed into the Select', () => {
  const select = <Select options={[{ value: 'm', display: 'moo' }, { value: 'b', display: 'boo' }]} onChange={() => {}} />;
  const wrapper = mount(select);
  expect(wrapper).toMatchSnapshot();
});
