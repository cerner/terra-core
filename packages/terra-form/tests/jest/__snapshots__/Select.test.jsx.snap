exports[`test should render a Select component with one option 1`] = `
<select
  className="select"
  name={null}
  required={false}>
  <option
    value="m">
    moo
  </option>
</select>
`;

exports[`test should render a Select when all the possible props are passed into it 1`] = `
<select
  aria-required="true"
  className="select"
  name="description"
  required={true}>
  <option
    value="m">
    moo
  </option>
</select>
`;

exports[`test should render a default Select component 1`] = `
<select
  className="select"
  name={null}
  required={false} />
`;

exports[`test should render as controlled when just the required fields and a onChange function is passed into the Select 1`] = `
<Select
  choices={null}
  name={null}
  onChange={[Function]}
  options={
    Array [
      Object {
        "display": "moo",
        "value": "m",
      },
      Object {
        "display": "boo",
        "value": "b",
      },
    ]
  }
  required={false}>
  <select
    className="select"
    name={null}
    onChange={[Function]}
    required={false}>
    <option
      value="m">
      moo
    </option>
    <option
      value="b">
      boo
    </option>
  </select>
</Select>
`;

exports[`test should render as uncontrolled when just the required fields and a default value is passed into the Select 1`] = `
<Select
  choices={null}
  defaultValue="b"
  name={null}
  options={
    Array [
      Object {
        "display": "moo",
        "value": "m",
      },
      Object {
        "display": "boo",
        "value": "b",
      },
    ]
  }
  required={false}>
  <select
    className="select"
    defaultValue="b"
    name={null}
    required={false}>
    <option
      value="m">
      moo
    </option>
    <option
      value="b">
      boo
    </option>
  </select>
</Select>
`;
